library/alloc/src/vec/mod.rs

```diff
@@ -3183,6 +3183,57 @@ pub fn from_elem_in<T: Clone, A: Allocator>(elem: T, n: usize, alloc: A) -> Vec<
     <T as SpecFromElem>::from_elem(elem, n, alloc)
 }
 
+/// # Safety
+///
+/// `elem` must be the result of some const expression.
+#[doc(hidden)]
+#[cfg(not(no_global_oom_handling))]
+#[unstable(feature = "vec_of_const_expr", issue = "none")]
+#[track_caller]
+pub unsafe fn from_const_elem<T>(elem: T, n: usize) -> Vec<T> {
+    // SAFETY: Caller has guaranteed `elem` being the result of some const expression.
+    unsafe { from_const_elem_in(elem, n, Global) }
+}
+
+/// # Safety
+///
+/// `elem` must be the result of some const expression.
+#[doc(hidden)]
+#[cfg(not(no_global_oom_handling))]
+#[unstable(feature = "vec_of_const_expr", issue = "none")]
+#[track_caller]
+unsafe fn from_const_elem_in<T, A: Allocator>(elem: T, n: usize, alloc: A) -> Vec<T, A> {
+    /// # Safety
+    ///
+    /// `value` must points to a valid `T` value that is the result of some const expression.
+    unsafe fn fill_const_value<T>(buffer: &mut [MaybeUninit<T>], value: *const T) {
+        for target in buffer {
+            // SAFETY: If `value` is the result of some const expression, we can make as many
+            // copies as needed.
+            unsafe { target.write(ptr::read(value)) };
+        }
+    }
+
+    // Avoid calling the destructor of `elem`.
```

Comment:
This raises the interesting question whether the destructor should be executed at all for `vec![const { ... }; 0]`. For 0-length arrays, our current behavior here is rather strange: https://github.com/rust-lang/rust/issues/79580.

Reply:
I tested the array of const expression case, the destructor is not called: <https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9566575c48c8b9928f861d8cd3da3871>. This is consistent with the current implementation.

Reply:
Test added.

Reply:
Sounds good!